query/workmanager.go

```diff
@@ -0,0 +1,447 @@
+			// remove it from our set of active workers.
+			onExit := make(chan struct{})
+			workers[peer.Addr()] = &activeWorker{
+				w:         r,
+				activeJob: nil,
+				onExit:    onExit,
+			}
+
+			w.cfg.Ranking.AddPeer(peer.Addr())
+
+			w.wg.Add(1)
+			go func() {
+				defer w.wg.Done()
+				defer close(onExit)
+
+				r.Run(peer, w.jobResults, w.quit)
+			}()
+
+		// A new result came back.
+		case result := <-w.jobResults:
+			log.Tracef("Result for job %v received. err=%v",
+				result.job.index, result.err)
+
+			// Delete the job from the worker's active job, such
+			// that the slot gets opened for more work.
+			r, ok := workers[result.peer.Addr()]
+			if ok {
+				r.activeJob = nil
+			}
+
+			// Get the index of this query's batch, and delete it
+			// from the map of current queries, since we don't have
+			// to track it anymore.
+			batchNum := currentQueries[result.job.index]
+			delete(currentQueries, result.job.index)
+			batch := currentBatches[batchNum]
+
+			switch {
+			case result.err != nil:
+				switch {
+				// If the query ended because it was canceled,
+				// drop it.
+				case result.err == ErrJobCanceled:
+					log.Tracef("Query(%d) was canceled "+
+						"before result was available",
+						result.job.index)
+
+					// If this is the first job in this
+					// batch that was canceled, forward the
+					// error on the batch's error channel.
+					// We do this since a cancellation
+					// applies to the whole batch.
+					if batch != nil {
+						batch.errChan <- result.err
+						delete(currentBatches, batchNum)
+
+						log.Debugf("Canceled batch %v",
+							batchNum)
+					}
+
+				// If the query ended with any other error, put
+				// it back into the work queue.
+				default:
+					log.Warnf("Query(%d) failed with "+
+						"error: %v. Rescheduling",
+						result.job.index, result.err)
+
+					// If it was a timeout, we dynamically
+					// increase it for the next attempt.
+					if result.err == ErrQueryTimeout {
+						newTimeout := result.job.timeout * 2
+						if newTimeout > maxQueryTimeout {
+							newTimeout = maxQueryTimeout
+						}
+						result.job.timeout = newTimeout
+					}
+
+					heap.Push(work, result.job)
+					currentQueries[result.job.index] = batchNum
+
+					// Punish the peer for the
+					// failed query.
+					w.cfg.Ranking.Punish(
+						result.peer.Addr(),
+					)
+
+				}
+
+			// Otherwise we update the status of the batch this
+			// query is a part of.
+			default:
+				// Get the batch and decrement the number of
+				// queries remaining.
+				if batch != nil {
+					batch.rem--
+					log.Tracef("Remaining jobs for batch "+
+						"%v: %v ", batchNum, batch.rem)
+
+					// If this was the last query in flight
+					// for this batch, we can notify that
+					// it finished, and delete it.
+					if batch.rem == 0 {
+						batch.errChan <- nil
+						delete(currentBatches, batchNum)
+						log.Tracef("Batch %v done",
+							batchNum)
```

Comment:
Ran into a deadlock:
```
1 @ 0x1032d60 0x1006edb 0x1006eb1 0x1006c95 0x152677f 0x1060bd1
#	0x152677e	github.com/lightninglabs/neutrino/query.(*WorkManager).workDispatcher+0xf8e	/Users/wilmer/git/neutrino/query/workmanager.go:378

1 @ 0x1032d60 0x1043350 0x104333b 0x1042fa2 0x107e954 0x155d423 0x155d3f4 0x1ad59ac 0x1a81e3d 0x1aeb662 0x103298e 0x1060bd1
#	0x1042fa1	sync.runtime_Semacquire+0x41						/usr/local/go/src/runtime/sema.go:56
#	0x107e953	sync.(*WaitGroup).Wait+0x63						/usr/local/go/src/sync/waitgroup.go:130
#	0x155d422	github.com/lightninglabs/neutrino/query.(*WorkManager).Stop+0xb2	/Users/wilmer/git/neutrino/query/workmanager.go:137
#	0x155d3f3	github.com/lightninglabs/neutrino.(*ChainService).Stop+0x83		/Users/wilmer/git/neutrino/neutrino.go:1548
#	0x1ad59ab	github.com/lightningnetwork/lnd.initNeutrinoBackend.func3+0x3b		/Users/wilmer/git/lnd/chainregistry.go:769
#	0x1a81e3c	github.com/lightningnetwork/lnd.Main+0x269c				/Users/wilmer/git/lnd/lnd.go:589
#	0x1aeb661	main.main+0x31								/Users/wilmer/git/lnd/cmd/lnd/main.go:14
#	0x103298d	runtime.main+0x21d							/usr/local/go/src/runtime/proc.go:203
```

I believe it's due to a missing continue here. I think we'll still want to reward the peer before doing so though.

Reply:
Good catch! Fixed: https://github.com/lightninglabs/neutrino/pull/179/commits/54b63afc0baa3250860d10b0081297efb56b004a